### PR TITLE
sdn: miscellaneous fixes after the CNI merge

### PR DESCRIPTION
**Pull request by dcbw**
The podManager must be started (so it can process requests) before we try to call Update on pods at startup if networking has changed.

Fixes: cf69a41bfed068adcd54b00bcf856120c0273c48
Fixes bug 1388856
Fixes bug 1389717

### CODE REVIEW

**Comment by dcbw** [comment]
@eparis @openshift/networking @knobunc

**Comment by danwinship** [comment]
LGTM. Did you mean to link to a bug rather than a commit id above?

**Comment by dcbw** [comment]
@danwinship meant the commit ID, though I could do the github issue too if you like.

**Comment by pravisankar** [comment]
LGTM

**Comment by eparis** [comment]
[test]

**Comment by eparis** [comment]
[testextended][extended:networking]

File: pkg/sdn/plugin/node.go [hunk 1]
```diff
@@ -207,6 +207,9 @@ func (node *OsdnNode) Start() error {
 	if err != nil {
 		return err
 	}
+	if err := node.podManager.Start(cniserver.CNIServerSocketPath); err != nil {
```
Comment (liggitt) [inline pkg/sdn/plugin/node.go]: Is the listener always ready once this returns?

Comment (dcbw) [inline pkg/sdn/plugin/node.go]: Yeah, processCNIRequests() will be running in a goroutine at this point, handling subsequent requests from UpdatePod and kubelet.

But there's one more part to this bug that I'll push as a second commit and get another round of LGTM.  @DirectXMan12 found the issues yesterday when running a combined process master and node that this PR should fix.

Comment (liggitt) [inline pkg/sdn/plugin/node.go]: is the listener established synchronously before the goroutine is spawned?

**Comment by dcbw** [comment]
@danwinship @pravisankar @knobunc PTAL thanks; one more commit to fix @DirectXMan12 issue from yesterday.

File: pkg/sdn/plugin/node.go (hunk 1 above)
Comment (liggitt) [inline pkg/sdn/plugin/node.go]: yup, found `l, err := net.Listen("unix", s.path)` in CNIServer#Start before the Serve goroutine is kicked off

Comment (liggitt) [inline pkg/sdn/plugin/node.go]: if `go s.Serve(l)` returns an error, nothing will log/handle/restart/exit... if it panics, the process will exit. Do we need to put handling around the Serve call inside the goroutine?

**Comment by dcbw** [comment]
@DirectXMan12 can you test these 2 commits with your setup?

**Comment by dcbw** [comment]
> if go s.Serve(l) returns an error, nothing will log/handle/restart/exit... if it panics, the process will exit. Do we need to put handling around the Serve call inside the goroutine?

@liggitt that's a good point, but the only reason Serve(l) will return an error is becuase setupHTTP2_Serve() returns an error, but that won't happen because shouldConfigureHTTP2ForServe() will return false due to lack of srv.TLSConfig, if I understand correctly.

**Comment by dcbw** [comment]
> if go s.Serve(l) returns an error, nothing will log/handle/restart/exit...

Good point, will look at it.

**Comment by danwinship** [comment]
Looks right. plugin.go:Init() could return an error without closing `kubeletInitReady`, but the node will eventually hit a glog.Fatal() in that case, so it doesn't matter.

**Comment by dcbw** [comment]
> if go s.Serve(l) returns an error, nothing will log/handle/restart/exit...
> 
> Good point, will look at it.

Handled; @danwinship and I looked at it yesterday and concluded that running Serve() forever was OK since we're using Unix domain sockets and thus most of the reasons Serve()/Accept() will return aren't really valid.

File: pkg/sdn/plugin/cniserver/cniserver.go [hunk 2]
```diff
@@ -141,7 +143,11 @@ func (s *CNIServer) Start(requestFunc cniRequestFunc) error {
 	}
 
 	s.SetKeepAlivesEnabled(false)
-	go s.Serve(l)
+	go utilwait.Forever(func() {
+		if err := s.Serve(l); err != nil {
+			glog.Warningf("CNI server Serve() failed: %v", err)
```
Comment (danwinship) [inline pkg/sdn/plugin/cniserver/cniserver.go]: based on other sdn code, I think the preferred behavior is to call (kubernetes/pkg/util/)runtime.HandleError()

**Comment by dcbw** [comment]
Extended networking test failure seems like a dind or docker issue:

```
Oct 31 11:16:25.304: INFO: At 2016-10-31 11:15:44 -0400 EDT - event for flow-checkl35gu: {kubelet nettest-node-1} FailedSync: Error syncing pod, skipping: failed to "StartContainer" for "flow-check" with RunContainerError: "runContainer: Error response from daemon: Cannot start container ec598cdedcc673560b770debe1b1d705e3fc0751182dcc82b0ba5c740ec846f1: Path /var/run/openvswitch/br0.mgmt is mounted on /run but it is not a shared or slave mount."
```

**Comment by knobunc** [comment]
@danw @pravisankar PTAL

**Comment by dcbw** [comment]
re-[testextended][extended:networking]

**Comment by dcbw** [comment]
re-[testextended][extended:networking] flake is https://github.com/openshift/origin/issues/11707

**Comment by marun** [comment]
Consider adding ``mount --make-shared /run`` after https://github.com/openshift/origin/blob/master/images/dind/dind-setup.sh#L44 and see if the test passes.

**Comment by dcbw** [comment]
re-[testextended][extended:networking] flake is #11707

**Comment by knobunc** [comment]
[merge]

**Comment by dcbw** [comment]
Extended net tests failed with:

SetupNetworkError: "Failed to setup network for pod \"service-wget_e2e-tests-net-services2-sr3qi(90a873e7-a08c-11e6-adad-0242d009400c)\" using network plugins \"cni\": CNI request failed with status 400: 'Failed to ensure that nat chain OUTPUT jumps to KUBE-HOSTPORTS: error checking rule: exit status 2: iptables v1.4.21: Couldn't load target `KUBE-HOSTPORTS':No such file or directory\n\nTry `iptables -h' or 'iptables --help' for more information.\n\n'; Skipping pod"

but that's not related to this PR, that's coming from pkg/kubelet/network/hostport/hostport.go:

```
	glog.V(4).Info("Ensuring kubelet hostport chains")
	// Ensure kubeHostportChain
	if _, err := h.iptables.EnsureChain(utiliptables.TableNAT, kubeHostportsChain); err != nil {
		return fmt.Errorf("Failed to ensure that %s chain %s exists: %v", utiliptables.TableNAT, kubeHostportsChain, err)
	}
```

**Comment by danwinship** [comment]
> but that's not related to this PR, that's coming from pkg/kubelet/network/hostport/hostport.go:

actually, it's coming from the EnsureRule() call a few lines down from that EnsureChain() call... but it looks like it's failing because the chain that it just called EnsureChain() on doesn't exist...

Is it possible for SyncHostports() to get called from multiple threads at once?

**Comment by dcbw** [comment]
> actually, it's coming from the EnsureRule() call a few lines down from that EnsureChain() call... but it looks like it's failing because the chain that it just called EnsureChain() on doesn't exist...
> 
> Is it possible for SyncHostports() to get called from multiple threads at once?

All the calls should be synchronized through the podManager since they are done from the setup/teardown liek the rest of the pod network operations.  They should also be synchronized by the iptables commands and the locking that the go iptables stuff uses, I think?

**Comment by abhgupta** [comment]
re-[test] The test failure is unrelated to the changes in this PR.

**Comment by danwinship** [comment]
[test] again so it will hopefully pass before the merge queue reaches it

**Pull request merged**